### PR TITLE
Support Mac Crosscompiler!

### PR DESCRIPTION
**Pull request by neilweidinger**
## Description
After a long, long time, I've finally set up the Mac crosscompiler :))

The `install-mac-cc.sh` script basically downloads a tarball with a copy of clang, beaglebone header files, and binutils compiled for the mac. The makefile then passes in the appropriate flags to and compiles the project using the tools downloaded in this tarball. The guide on how to use this is already up on the [wiki here](https://github.com/Hyp-ed/hyped-2020/wiki/Using-the-Mac-crosscompiler-for-the-BBB).

### CODE REVIEW

File: src/utils/io/spi.cpp [hunk 1]
```diff
@@ -28,35 +28,35 @@
 #include <unistd.h>
 #include <sys/mman.h>
 
-#ifndef WIN
-#include <linux/spi/spidev.h>
+#if LINUX
+    #include <linux/spi/spidev.h>
```
Comment (mkristien) [inline src/utils/io/spi.cpp]: Preprocessor directives should not create indentation levels, as per our [styleguide](https://hyp-ed.github.io/styleguide/#Preprocessor_Directives).

File: Makefile [hunk 2]
```diff
@@ -28,28 +28,33 @@ GITHOOKS=.git/hooks
...
+					  --sysroot=mac-crosscompiler/sysroot \
+					  -isysroot mac-crosscompiler/sysroot \
+					  -Imac-crosscompiler/sysroot/usr/include/c++/6.3.0 \
+					  -Imac-crosscompiler/sysroot/usr/include/arm-linux-gnueabihf/c++/6.3.0 \
+					  --gcc-toolchain=mac-crosscompiler/prebuilt/bin
+	LFLAGS:= $(LFLAGS) --target=arm-linux-gnueabihf -L$(COMPILER_PATH) --sysroot=mac-crosscompiler/sysroot
 else
+    $(info cross-compiling using Linux host)
 	CC:=hyped-cross-g++
-	CFLAGS:=$(CFLAGS) -DARCH_32
```
Comment (mkristien) [inline Makefile:44]: Why is this removed?

File: Makefile [hunk 3]
```diff
@@ -28,28 +28,33 @@ GITHOOKS=.git/hooks
...
 	ARCH=$(shell uname -m)
 	ifneq (,$(findstring 64,$(ARCH)))
 		CFLAGS:=$(CFLAGS) -DARCH_64
 	endif
+else ifeq ($(UNAME), Darwin)
+    $(info cross-compiling using Mac master race host)
+	CC:=mac-crosscompiler/prebuilt/bin/clang++
+	export COMPILER_PATH = mac-crosscompiler/sysroot/usr/lib/gcc/arm-linux-gnueabihf/6.3.0/
+	CFLAGS:=$(CFLAGS) --target=arm-linux-gnueabihf \
+					  --sysroot=mac-crosscompiler/sysroot \
```
Comment (mkristien) [inline Makefile:40]: Is this additional stuff necessary? Is there no g++ cross compiler? Could you perhaps symlink this to `hyped-cross-g++` to make it consistent with other cross-compilers?

**Review by mkristien — CHANGES_REQUESTED** [review]
There are many Makefile changes, I am worried about compatibility on other platforms.

This relates to the makefile refactoring [PR](https://github.com/Hyp-ed/hyped-2020/pull/45). We need to merge those at some point. Now sure how to approach this, but you might need to redo this setup there as well.

File: src/utils/io/spi.cpp (hunk 1 above)
Comment (neilweidinger) [reply src/utils/io/spi.cpp]: Thanks for pointing that out, will fix, didn't even know that was in our style guide.

File: Makefile (hunk 2 above)
Comment (neilweidinger) [reply Makefile:44]: Removed this because `-DARCH_32` wasn't being used anywhere, unlike `-DARCH_64`.

File: Makefile (hunk 3 above)
Comment (neilweidinger) [reply Makefile:40]: I tried looking into a g++ cross compiler, but setting it up seems very complex (that isn't the precompiled linux one) and using clang is relatively simple as it's natively a crosscompiler (some flags just have to be specified as you can see). I was thinking of updating the linux crosscompiler to use clang as well in the future.

File: src/utils/io/spi.cpp (hunk 1 above)
Comment (neilweidinger) [reply src/utils/io/spi.cpp]: Actually can we just remove this from our style guide? The [reasons](https://stackoverflow.com/a/789156/7162381) for not indenting preprocessor directives seem to be quite outdated and don't even apply to us (unless there's something I'm overlooking). Imo it's more readable when indented.

Comment (mkristien) [reply src/utils/io/spi.cpp]: Please, raise an issue on the styleguide repo and we will vote :)

File: Makefile (hunk 3 above)
Comment (mkristien) [reply Makefile:40]: What does "it's natively a crosscompiler" mean?

File: install-mac-cc.sh [hunk 4]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
```
Comment (mkristien) [inline install-mac-cc.sh]: Do you think this file could be hidden in some `utils/` folder?

**Comment by mkristien** [comment]
Btw, thanks for the docs on the wiki.

Please, address the concerns about platform portability (have you tried this on BBB?) and the interaction with the makefile-refactoring PR.

File: Makefile (hunk 3 above)
Comment (neilweidinger) [reply Makefile:40]: Instead of having to compile gcc for a specific host/target combination, clang [supports compiling to any target](https://clang.llvm.org/docs/CrossCompilation.html) out of the box.

File: src/utils/io/spi.cpp (hunk 1 above)
Comment (neilweidinger) [reply src/utils/io/spi.cpp]: Haha fair, I'll put up an issue after this pr.

File: install-mac-cc.sh (hunk 4 above)
Comment (neilweidinger) [reply install-mac-cc.sh]: Currently the script is dependent on which directory you're in when you run it, as it's designed to be run while you're in `hyped-2020` (uses relative pathnames). I would prefer using absolute pathnames, but I'm not entirely sure how to do this in a script. I'm a bit worried that if it was in a `utils/` directory people might run `./utils/install-mac-cc.sh` or `cd` into `utils/` first and then run `./install-mac-cc.sh`, and that would result in different things.

**Comment by neilweidinger** [comment]
Of course, platform portability should be the highest priority. Do you mean if I've actually tried running the crosscompiled binary on a BBB, or if the makefile still works when compiling directly on the BBB?

I haven't tried running it with the makefile-refactoring PR yet.

File: install-mac-cc.sh (hunk 4 above)
Comment (neilweidinger) [reply install-mac-cc.sh]: Never mind, [this answer on SO](https://stackoverflow.com/a/246128/7162381) seems to do the trick in finding an absolute pathname of the script, will refactor and move it to `utils/`.

File: Makefile (hunk 3 above)
Comment (mkristien) [reply Makefile:40]: Previously, setting up cross compiler on Windows or Linux never involved compiling the gcc from scratch. We just installed one online, similarly how this PR installs the clang.

**Comment by mkristien** [comment]
Actually, both. Have you verified the cross compiled hyped runs on BBB?

I guess, running anything while not cross compiling would still work. My worry is if running cross compilation on Linux/Windows still produces the right thing.

One issue I have found so far is the `if LINUX` in `utils/io/spi.cpp`. When cross compiling, this should be defined regardless of the host platform, as the code is compiled as if on Linux.

File: Makefile (hunk 3 above)
Comment (neilweidinger) [reply Makefile:40]: Yes, but that's only because we were lucky someone had already compiled gcc targeting the raspberry pi for x86 hosts, we simply downloaded those files and used that. This uses a precompiled version of clang (available on their website to download) for mac hosts, but had to compile binutils from source unfortunately. I also copied over the beaglebone include and library files.

File: src/utils/io/can.cpp [hunk 5]
```diff
@@ -26,23 +26,23 @@
 #include <sys/socket.h>
 #include <net/if.h>
 
-#ifndef WIN
-#include <linux/can.h>
+#if LINUX
```
Comment (mkristien) [inline src/utils/io/can.cpp:29]: Please define this when cross compiling. Furthermore, `#if <VAR>` checks the value is true (non-zero), while `#ifdef <VAR>` checks the variable exists. In other words, the first one requires `-DVAR=1` compiler flag, while the second requires only `-DVAR`.

Comment (neilweidinger) [reply src/utils/io/can.cpp:29]: `-DVAR` [without anything specified](https://gcc.gnu.org/onlinedocs/gcc/Preprocessor-Options.html#Preprocessor-Options) defaults to `1`, so the `#if` should still catch it.

**Comment by neilweidinger** [comment]
I've run the crosscompiled binary on the beaglebone and there don't seem to be any issues.

File: Makefile (hunk 3 above)
Comment (mkristien) [reply Makefile:40]: We used the same for MAC, how about this one? https://s3.amazonaws.com/jaredwolff/xtools-2016-09-01.dmg


I actually do not mind either way, as I will probably never use this. It just seems like it would be much better if every platform used almost the same setup, and then the same Makefile script relevant to cross-compilation. This way, just be prepared I will give no support for cross-compilation issues to MAC users. I will also take every opportunity to make fun of MAC users is things do not work, as usual :)

Comment (neilweidinger) [reply Makefile:40]: That's the file I meant by someone precompiled gcc. I agree it'd be nice if the same crosscompiler tools could be used for mac and linux, but at the end of the day there are inevitable differences between the two and if not now, then later there might be slight changes required to one platform to support it while keeping things the same for the other anyways.

I'll take responsibility for maintaining and supporting the mac crosscompiler :)

File: src/utils/io/can.cpp (hunk 5 above)
Comment (mkristien) [reply src/utils/io/can.cpp:29]: good to know

**Comment by mkristien** [comment]
Do you see any compilation warning when compiling `utils/io/spi.cpp`? I think you would not have `#if LINUX` when cross compiling, which would probably result in some runtime issues when trying to actually use spi?

**Comment by neilweidinger** [comment]
Thank you, good catch. I've updated the makefile so that the mac crosscompiler uses the linux SPI and CAN include files.

**Review by mkristien — APPROVED** [review]
